ports-of-call/math_utils.hpp

```diff
@@ -0,0 +1,99 @@
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef PORTS_OF_CALL_MATH_UTILS_HPP_
+#define PORTS_OF_CALL_MATH_UTILS_HPP_
+
+#include <cmath>
+#include <limits>
+#include <ports-of-call/portability.hpp>
+#include <ports-of-call/robust_utils.hpp>
+
+namespace PortsOfCall {
+namespace Math {
+
+// Faster implementation of std::pow for arithmetic bases and non-negative integer
+// exponents.  For sufficiently large integer powers std::pow may be faster, but testing
+// indicates that the following implementation is significantly faster (roughly a factor
+// of two or better) up to powers of at least 100.
+template <typename BaseT, typename ExponentT,
+          typename std::enable_if<std::is_arithmetic_v<std::decay_t<BaseT>> &&
+                                  std::is_integral_v<std::decay_t<ExponentT>>>::type * =
+              nullptr>
+[[gnu::always_inline]] PORTABLE_INLINE_FUNCTION constexpr auto power(BaseT base,
+                                                                     ExponentT exponent) {
+  using std::pow;
+  using PowT = decltype(pow(base, exponent));
+  if (!Robust::check_nonnegative(exponent) || exponent > ExponentT{100}) {
+    return pow(base, exponent);
+  }
+  PowT result = PowT{1};
+  while (true) {
+    if (exponent & 1) result *= base; // Multiply if the remaining exponent is odd
+    exponent >>= 1;                   // Right-shift the exponent (divide by 2)
+    if (!exponent) break;             // If the remaining exponent is zero, we are done
+    base *= base;                     // We halved the exponent, so square the base
+  }
+  return result;
+}
+// Faster implementation of std::pow() for non-negative arithmetic bases and
+// floating-point exponents
+template <typename BaseT, typename ExponentT,
+          typename std::enable_if<std::is_arithmetic_v<std::decay_t<BaseT>> &&
+                                  std::is_floating_point_v<std::decay_t<ExponentT>>>::type
+              * = nullptr>
+[[gnu::always_inline]] PORTABLE_INLINE_FUNCTION constexpr auto
+power(BaseT const &base, ExponentT const &exponent) {
+  using std::exp;
+  using std::log;
+  using std::pow;
+  if (!Robust::check_nonnegative(base)) {
+    return pow(base, exponent);
+  }
+  return exponent == ExponentT{0} ? BaseT{1} // Enforcing base^0=1 (including 0^0=1)
+         : base == BaseT{0}       ? BaseT{0}
+                                  : exp(exponent * log(base));
+}
+// Overload for non-arithmetic bases or exponents
+template <typename BaseT, typename ExponentT,
+          typename std::enable_if<not std::is_arithmetic_v<std::decay_t<BaseT>> ||
+                                  not std::is_arithmetic_v<std::decay_t<ExponentT>>>::type
+              * = nullptr>
+[[gnu::always_inline]] PORTABLE_INLINE_FUNCTION constexpr auto
+power(BaseT const &base, ExponentT const &exponent) {
+  using std::pow;
+  return pow(base, exponent);
```

Comment:
seems a bit silly to use a `using` statement for a one-liner
```suggestion
  return std::pow(base, exponent);
```

Reply:
Having worked a bunch with CJ, I'm pretty sure this came from him and is designed for ADL.  I had the same thought as you, but wasn't sure if we wanted to kill ADL.  Given that this will pretty much always be used for things like `int`, `double`, and other built-in POD types, it's probably a non-issue to drop ADL.  Having said that: Do you still prefer I make the recommended change?  If so, there are similar things elsewhere, should I change them too?

Reply:
I don't feel that strongly about it. If you're concerned we'll need non-standard types downstream for, e.g., this function to be called in EAP or Singe, we can leave it in.

Personally, I find aggressively type agnostic coding to be a big hit to readability and prefer being more explicit. But this is a shared capability and we can make compromises, especially since these functions in particular were obviously designed to be flexible as far as input type.

Reply:
Of all my comments, the thing I feel most strongly about is `PORTABLE_FORCEINLINE_FUNCTION` as that should be more flexible than manually inserting compiler specific directives about inlining.

Reply:
I was happy to make the `PORTABLE_FORCEINLINE_FUNCTION` change.  I just forgot that existed and was copying code from somewhere else.

The change regarding the `using` statements has been made.

Reply:
Sounds good!